### PR TITLE
Add self-hosted Grok OAuth connector

### PR DESCRIPTION
**Pull request by XiyaoWang0519**
## Summary

- add an optional self-hosted, single-owner OAuth 2.1 endpoint at `/grok/mcp/` for Grok Bot Custom Connectors
- preserve the existing `/mcp/` bearer plus `X-Agent-User-Id` authentication used by OpenClaw and Hermes
- add owner-secret consent, Argon2id verification, PKCE S256, dynamic client registration, rotating refresh tokens, revocation, encrypted SQLite persistence, bounded registration/audit storage, and request/rate limits
- add Grok connector documentation, owner-secret hash tooling, regression coverage, and a no-call local smoke flow

## Risk and rollout

- OAuth is disabled by default and adds no behavior to existing deployments until explicitly configured
- this changes the public authentication surface and adds SQLite tables, but does not change the call state machine, explicit call confirmation, webhook verification, or deployment lease
- live Grok connector verification remains pending until an approved deployment; Grok Voice remains out of scope
- no real phone call or production deployment was performed for this PR

## Validation

- `uv sync --all-groups --frozen`
- `uv run ruff format --check app tests scripts`
- `uv run ruff check app tests scripts`
- `uv run mypy app`
- clean worktree: `396 passed`, `87.85%` coverage on Python 3.12
- active checkout: `398 passed`, `87.36%` coverage on Python 3.13; the two extra tests come from an unrelated preserved duplicate test file
- `uv run pre-commit run --all-files`
- `bash scripts/live_smoke.sh` passed OAuth discovery, DCR, consent, token exchange, exactly seven OAuth tools, exactly seven legacy tools, and `prepare_phone_call`; it never invoked `start_phone_call`

### CODE REVIEW

File: app/routes/grok_oauth.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    form = await request.form()
+    transaction_id = str(form.get("tx") or "")
+    action = str(form.get("action") or "")
+    csrf_token = str(form.get("csrf_token") or "")
+    owner_secret = str(form.get("owner_secret") or "")
+    transaction = await provider.load_transaction(transaction_id)
+    if transaction is None or not provider.verify_csrf(transaction, csrf_token):
+        provider.record_failed_attempt(key)
+        return secure_html(consent_error_page(GENERIC_FAILURE), status_code=400)
```
[the remark this answers is not in the document]
Comment (XiyaoWang0519) [reply app/routes/grok_oauth.py:55]: Fixed. Invalid or expired `tx` and CSRF mismatches now return 400 without calling `record_failed_attempt`, so anonymous garbage POSTs no longer increment the account-wide owner-secret limiter. Wrong owner secrets on a valid transaction still count toward the 5/15-minute lockout.

File: app/grok_oauth/provider.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+            "scopes": scopes,
+            "resource": resource,
+            "code_challenge": params.code_challenge,
+            "csrf_token": csrf_token,
+        }
+        await self._store().oauth_create_transaction(
+            transaction_id=transaction_id,
+            csrf_hash=self.hash_secret_material(csrf_token),
+            ciphertext=self.encrypt_payload(payload),
+            expires_at=expires_at.isoformat(),
```
[the remark this answers is not in the document]
Comment (XiyaoWang0519) [reply app/grok_oauth/provider.py]: Fixed. `/authorize` now purges expired or consumed `oauth_auth_transactions` rows before insert and rejects new requests once 64 outstanding transactions exist globally or 8 exist for the same client.

File: app/grok_oauth/provider.py [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+            revocation_options=RevocationOptions(enabled=True),
+            required_scopes=[GROK_OAUTH_SCOPE],
+        )
+        self._settings = settings
+        self.resource = resource
+        owner_hash = Settings.reveal(settings.grok_mcp_oauth_owner_secret_hash)
+        signing_material = Settings.reveal(settings.grok_mcp_oauth_signing_key)
+        storage_material = Settings.reveal(settings.grok_mcp_oauth_storage_encryption_key)
+        self._owner_secret_hash = owner_hash
+        self._fernet = derive_storage_fernet(storage_material, salt=issuer)
```
[the remark this answers is not in the document]
Comment (XiyaoWang0519) [reply app/grok_oauth/provider.py]: Fixed. Storage Fernet and MAC keys are derived from a stable deployment salt plus `GROK_MCP_OAUTH_STORAGE_ENCRYPTION_KEY`, not `PUBLIC_BASE_URL`. Changing the tunnel URL no longer invalidates the sentinel or persisted ciphertext. Access-token signing remains issuer-bound.
